Request scope during token exchange

NobodysNightmare opened the pull request:
# Ticket
https://community.openproject.org/wp/64121

# What are you trying to accomplish?
Allowing to specify scopes requested during Token Exchange. This is effectively required with the latest Keycloak version, because the scope parameter allows to ask for any `Optional` scopes, thus it allows to extend the scope beyond the scope of the original token. On the other hand the `audience` parameter can only be used to reduce the token's existing audiences to the one requested. This means exchanging a login token with audience `A` for a token with audience `B` requires to add a scope that will result in an extension of the audience, while at the same time filtering out all other audiences using the `audience` parameter.

To make the forms more consistent, the framing of "storage audience" has been changed to "token exchange", where the audience is just one aspect of it.

## Screenshots

![image](https://github.com/user-attachments/assets/b504db44-60c3-43af-bdf5-ef1d705287ff)

![image](https://github.com/user-attachments/assets/f6ff5301-ba5e-4e3a-8159-8cd0f161fb90)

NobodysNightmare committed:
Allow to pass a scope during token exchange

This is the most basic/immediate implementation of this.
It's now possible to specify scopes that will be requested during
token exchange for storage SSO.

An immediate next step would probably be to rephrase the former
"audience" UI to be more inclusively phrased with regards to containing
settings for token exchange in general.

In the longer run we should also rethink the "backend" code of token exchange:

* The scope is so far not considered at all when figuring out whether we already
  have a matching token, we only compare the audience for that
* We should probably not read audiences of the JWT anymore and only decide between
  performing token exchange (requiring an audience) or not doing it. The currently supported
  use case of configuring an audience that's then expected in the login token does not really
  add any value.
* it's again a question how many authentication details should be stored on a storage
  instead of being stored somewhere away from the storage

NobodysNightmare committed:
Reword storage audience component towards token exchange

Making it more clear to the user that this component does not
only configure an audience, but rather that it configures the
token exchange (or lack thereof).

NobodysNightmare committed:
Use expect syntax for strong parameters and sort actions

Making Rubocop happy, one file at a time...

mereghost commented:
Seems reasonable. The scope can have the same name as the audience, should we make this a configuration requirement? This way we wouldn't need to store both.

NobodysNightmare commented:
> Seems reasonable. The scope can have the same name as the audience, should we make this a configuration requirement? This way we wouldn't need to store both.

I was thinking about this, but it feels weird/ugly/wrong to me to force a conventional scope name, when we could leave that decision up to the admin. E.g. I'd have thought to call it `obtain-nextcloud-token`, but there can also be an argument to call it `nextcloud`. But we should we _force_ an admin to choose a specific name in their deployment?

Another consideration: While we can force it for Keycloak and we have a specific reason to request a scope in the Keycloak case, maybe for other IDPs different scopes are needed as well. So I feel that by allowing admins to add scopes here, we again make this integration more generically usable. I.e. we increase the chances that an integration will work somewhere else as well.

NobodysNightmare committed:
Incorporate design feedback

Kharonus commented on modules/openid_connect/app/services/openid_connect/user_tokens/token_request.rb:
```diff
@@ -43,13 +43,16 @@ def refresh(refresh_token)
+          subject_token_type: OpenProject::OpenIDConnect::ACCESS_TOKEN_TYPE,
+          audience:
+        }
+        parameters[:scope] = scope unless scope.nil?
```
❓ why this extra route? What happens if the parameters contains `scope: nil`? And is calling `request_token(form: parameters.compact)` an option?

NobodysNightmare replied:
I essentially wanted to allow the full bandwidth of client side behaviour here: Passing a scope (any string), passing an empty scope (empty string) or not passing a scope at all.

Whether any IDP reacts badly to an empty scope parameter I can't say for sure right now.

`parameters.compact` would have been an option theoretically, but right now I only wanted to support omission of the `scope` parameter, but not (for example) of the `audience` parameter.

In the end this whole branch is a bunch of "I have no clue if I am going to need it", but it felt more correct to implement it like this, rather than always sending a scope no matter what.

Kharonus commented on modules/storages/app/models/storages/nextcloud_storage.rb:
```diff
@@ -49,6 +49,7 @@ class NextcloudStorage < Storage
     store_attribute :provider_fields, :group_folder, :string
     store_attribute :provider_fields, :authentication_method, :string, default: "two_way_oauth2"
     store_attribute :provider_fields, :storage_audience, :string
+    store_attribute :provider_fields, :storage_scope, :string
```
🟡 as an attribute accessor, I'm no big fan of `storage_scope` and `storage_audience`, because we are just repeating the object itself, leading to weird access like `storage.storage_scope`, while it could be simply `storage.scope` or `storage.[oidc|sso]_scope`.

Kharonus replied:
Just to be clear, in the user-facing form itself I see no problem, as we have those wizard steps, putting the input form fields more away from the "Storage" as a thing.

NobodysNightmare replied:
I guess that's fair. I am not going to work against `storage_audience` in this PR. But I can rename the scope.

I guess `scope` is too unspecific for me. `sso_scope` is closer to what I like. What do you think about `token_exchange_scope`? It makes clear what we use it for and also silenty screams "can you put this configuration somewhere else?", which is something that I still have in my mind for a future improvement (without any clear opinion on how exactly to change it, yet).

NobodysNightmare replied:
we agreed on `token_exchange_scope`

Kharonus reviewed (approved):
Looks good in general 🫡 

I would have some minor requests, but those are subjective ones and should be no blockers.

NobodysNightmare committed:
Reword: storage_scope to token_exchange_scope

It avoids duplicating the class name in the front of
its own attribute. The new name is also clearer on the
fact where this is used (because it only affects certain
aspects of the storage).
